src/rust/src/backend/kdf.rs

```diff
@@ -447,12 +449,239 @@ impl Argon2id {
+        key_material: CffiBuf<'_>,
+        expected_key: CffiBuf<'_>,
+    ) -> CryptographyResult<()> {
+        let actual = self.derive(py, key_material)?;
+        let actual_bytes = actual.as_bytes();
+        let expected_bytes = expected_key.as_bytes();
+
+        if actual_bytes.len() != expected_bytes.len()
+            || !openssl::memcmp::eq(actual_bytes, expected_bytes)
+        {
+            return Err(CryptographyError::from(exceptions::InvalidKey::new_err(
+                "Keys do not match.",
+            )));
+        }
+
+        Ok(())
+    }
+}
+
+#[pyo3::pyclass(
+    module = "cryptography.hazmat.primitives.kdf.hkdf",
+    name = "HKDFExpand"
+)]
+struct HkdfExpand {
+    algorithm: pyo3::Py<pyo3::PyAny>,
+    info: pyo3::Py<pyo3::types::PyBytes>,
+    length: usize,
+    used: bool,
+}
+
+#[pyo3::pymethods]
+impl HkdfExpand {
+    #[new]
+    #[pyo3(signature = (algorithm, length, info, backend=None))]
+    fn new(
+        py: pyo3::Python<'_>,
+        algorithm: pyo3::Py<pyo3::PyAny>,
+        length: usize,
+        info: Option<pyo3::Py<pyo3::types::PyBytes>>,
+        backend: Option<pyo3::Bound<'_, pyo3::PyAny>>,
+    ) -> CryptographyResult<Self> {
+        _ = backend;
+
+        let algorithm_bound = algorithm.bind(py);
+        let digest_size = algorithm_bound
+            .getattr(pyo3::intern!(py, "digest_size"))?
+            .extract::<usize>()?;
+
+        let max_length = 255 * digest_size;
+        if length > max_length {
+            return Err(CryptographyError::from(
+                pyo3::exceptions::PyValueError::new_err(format!(
+                    "Cannot derive keys larger than {} octets.",
+                    max_length
+                )),
+            ));
+        }
+
+        let info = if let Some(info) = info {
+            info
+        } else {
+            pyo3::types::PyBytes::new(py, b"").into()
+        };
+
+        Ok(HkdfExpand {
+            algorithm,
+            info,
+            length,
+            used: false,
+        })
+    }
+
+    fn derive<'p>(
+        &mut self,
+        py: pyo3::Python<'p>,
+        key_material: CffiBuf<'_>,
+    ) -> CryptographyResult<pyo3::Bound<'p, pyo3::types::PyBytes>> {
+        if self.used {
+            return Err(exceptions::already_finalized_error());
+        }
+        self.used = true;
+
+        let algorithm_bound = self.algorithm.bind(py);
+
+        let mut output = Vec::new();
+        let mut counter = 1u8;
+        let mut previous_output = Vec::new();
+
+        let h_prime = Hmac::new_bytes(py, key_material.as_bytes(), algorithm_bound)?;
+        while output.len() < self.length {
+            let mut h = h_prime.copy(py)?;
+            h.update_bytes(&previous_output)?;
+            h.update_bytes(self.info.as_bytes(py))?;
+            h.update_bytes(&[counter])?;
+
+            let block = h.finalize(py)?;
+            let block_bytes = block.as_bytes();
+            previous_output = block_bytes.to_vec();
```

Comment:
We don't want to convert to vec, we want to just have each iteration slice the back off of previous output